js/general_patterns.md

```diff
@@ -0,0 +1,210 @@
+# General Patterns
+
+This document provides an overview of the general architecture and programming
+patterns that will be used throughout the JavaScript API. These patterns have
+been chosen because of the flexibility they provide in integrating with native
+APIs and ease to build new interfaces from, and for their ability to create
+performant and more optimizable code.
+
+
+
+### Asynchronous Calls
+
+The asynchronous callback pattern is very similar to `setTimeout()`. With the
+exception that required arguments must always go first. After the callback, N
+arguments can be passed that will propagate to the callback.
+
+Example of pattern:
+```js
+asyncCall(<arguments>, callback[, ...args])
+```
+
+`<arguments>` are the arguments required by the function call.
+
+`callback` is the function that is executed when the asynchronous operation is
+complete.
+
+`...args` are the additional values that will propagate to the callback.
```

Comment:
I guess what I was getting at is that it just looks a little strange to see arguments, callback, more arguments, when people are so conditioned to callbacks coming last. I definitely see the utility of doing it though.

I would absolutely love it if all parameters were required and each function had a single signature to avoid situations like [listen()](https://github.com/nodejs/io.js/blob/master/lib/net.js#L1290-L1349). I think that would be reasonable for a low level API.